### DISCUSSION TITLE
Failed to alloc highmem ... error when booting Linux kernel

### DISCUSSION THREAD

**Discussion by masahir0y** [ipxe/ipxe#514] [General]
When I try to boot the Linux kernel, I observe this error:
 
    EFI stub: ERROR: Failed to alloc highmem for files

I observed it on CentOS 7.9 and CentOS 8.4.

This error is printed by the kernel:

    https://github.com/torvalds/linux/blob/v4.18/drivers/firmware/efi/libstub/efi-stub-helper.c#L591

But, I do not think the root cause exists in the kernel because the issue happens depending on the ipxe version.

I use ipxe.iso created by:

    make EMBED=<my-script-path> bin-x86_64-efi/ipxe.iso

The condition to reproduce the issue is somewhat complex.

I see it depending on the combination of the following three:

[1] depends on ipxe version

   'git bisect' points out the first bad commit is
   988d2c13cdf0 ("[efi] Use segment and bus number to identify PCI root bridge I/O protocol")

   If I revert it, this issue disappears.

[2] depends on the hardware

   I tested 3 server vendors.

   I do not observe this error on Quanta or Dell servers.
   I see it happening on an SuperMicro servers.

[3] depends on the commands you run from ipxe

   If I slightly modify my ipxe script, this issue disappears.

Here are two scripts, ok.ipxe and ng.ipxe
The difference between them is just one line; the initial value of the iterator (`set i:int32 `)

ok.ipxe:

    #!ipxe

    # parameters
    set host_ip:ipv6  240b:c0e0:103:5479:b422:2:1:0
    set host_gw:ipv6  240b:c0e0:103:5479::ffff
    set isolinux_url  http://[REDACTED]@[240b:c0e0:102:54e9:c06b:2:0:1]/path/to/data/dir

    set i:int32 1
    :loop isset ${net${i}/mac} || goto shell
      ifopen net${i}
      set net${i}/ip6:ipv6 ${host_ip}
      set net${i}/len6:int8 64
      set net${i}/gateway6:ipv6 ${host_gw}
      sleep 5
      kernel ${isolinux_url}/vmlinuz && goto bootkernel || ifclose net${i}
      inc i && goto loop

    :shell
     echo Could not fetch kernel from any device. Falling back to shell
     shell

    :bootkernel

      initrd ${isolinux_url}/initrd.img

      boot vmlinuz initrd=initrd.img console=ttyS1,115200

ng.ipxe:

    #!ipxe

    # parameters
    set host_ip:ipv6  240b:c0e0:103:5479:b422:2:1:0
    set host_gw:ipv6  240b:c0e0:103:5479::ffff
    set isolinux_url  http://[REDACTED]@[240b:c0e0:102:54e9:c06b:2:0:1]/path/to/data/dir

    set i:int32 0
    :loop isset ${net${i}/mac} || goto shell
      ifopen net${i}
      set net${i}/ip6:ipv6 ${host_ip}
      set net${i}/len6:int8 64
      set net${i}/gateway6:ipv6 ${host_gw}
      sleep 5
      kernel ${isolinux_url}/vmlinuz && goto bootkernel || ifclose net${i}
      inc i && goto loop

    :shell
     echo Could not fetch kernel from any device. Falling back to shell
     shell

    :bootkernel

      initrd ${isolinux_url}/initrd.img

      boot vmlinuz initrd=initrd.img console=ttyS1,115200

Running ok.ipxe on ipxe commit 2265a65 successfully boots the kernel:

    iPXE 1.0.0+ -- Open Source Network Boot Firmware -- http://ipxe.org
    Features: DNS HTTP iSCSI TFTP SRP AoE EFI Menu
    http://[240b:c0e0:102:54e9:c06b:2:0:1]/path/to/data/dir/vmlinuz.................. Connection timed out (http://ipxe.org/4c0a6092)
    http://[240b:c0e0:102:54e9:c06b:2:0:1]/path/to/data/dir/vmlinuz... ok
    http://[240b:c0e0:102:54e9:c06b:2:0:1]/path/to/data/dir/initrd.img... ok
    [    0.000000] Linux version 4.18.0-305.3.1.el8.x86_64 (mockbuild@kbuilder.bsys.centos.org) (gcc version 8.4.1 20200928 (Red Hat 8.4.1-1) (GCC)) #1 SMP Tue Jun 1 16:14:33 UTC 2021
    [    0.000000] Command line: vmlinuz initrd=initrd.img console=ttyS1,115200

Running ng.ipxe on ipxe commit 2265a65 fails:

    iPXE 1.0.0+ -- Open Source Network Boot Firmware -- http://ipxe.org
    Features: DNS HTTP iSCSI TFTP SRP AoE EFI Menu
    http://[240b:c0e0:102:54e9:c06b:2:0:1]/path/to/data/dir/vmlinuz.................. Connection timed out (http://ipxe.org/4c0a6092)
    http://[240b:c0e0:102:54e9:c06b:2:0:1]/path/to/data/dir/vmlinuz.................. Connection timed out (http://ipxe.org/4c0a6092)
    http://[240b:c0e0:102:54e9:c06b:2:0:1]/path/to/data/dir/vmlinuz... ok 
    http://[240b:c0e0:102:54e9:c06b:2:0:1]/path/to/data/dir/initrd.img... ok
    EFI stub: ERROR: Failed to alloc highmem for files
    Trying to load files to higher address

Running ng.ipxe on ipxe commit 2265a65 + revert 988d2c1 boots the kernel

    iPXE 1.0.0+ -- Open Source Network Boot Firmware -- http://ipxe.org
    Features: DNS HTTP iSCSI TFTP SRP AoE EFI Menu
    http://[240b:c0e0:102:54e9:c06b:2:0:1]/path/to/data/dir/vmlinuz.................. Connection timed out (http://ipxe.org/4c0a6092)
    http://[240b:c0e0:102:54e9:c06b:2:0:1]/path/to/data/dir/vmlinuz.................. Connection timed out (http://ipxe.org/4c0a6092)
    http://[240b:c0e0:102:54e9:c06b:2:0:1]/path/to/data/dir/vmlinuz.......... ok          
    http://[240b:c0e0:102:54e9:c06b:2:0:1]/path/to/data/dir/initrd.img... ok
    [    0.000000] Linux version 4.18.0-305.3.1.el8.x86_64 (mockbuild@kbuilder.bsys.centos.org) (gcc version 8.4.1 20200928 (Red Hat 8.4.1-1) (GCC)) #1 SMP Tue Jun 1 16:14:33 UTC 2021
    [    0.000000] Command line: vmlinuz initrd=initrd.img console=ttyS1,115200

**Comment by YaronOva** [comment]
Hi Masahir0y,
did you ever found the root cause of that issue?
why reverting this commit solve this issue? (commit https://github.com/ipxe/ipxe/commit/2265a65191d76ce367913a61c97752ab88ab1a59 + revert https://github.com/ipxe/ipxe/commit/988d2c13cdf0f0b4140685af35ced70ac5b3283c boots the kernel)
